### PR TITLE
chore(vrl): improve performance of VrlTarget::into_events

### PR DESCRIPTION
**Pull request by JeanMertz**
Another hotspot in the flamegraphs that might be improved by this change.

This removes a bunch of heap-allocated boxed trait objects for dispatching using an enum.

### CODE REVIEW

File: lib/vector-core/src/event/vrl_target.rs [hunk 1]
```diff
@@ -29,6 +29,56 @@ pub enum VrlTarget {
...
+pub enum TargetEvents {
+    One(Event),
+    Logs(TargetIter<LogEvent>),
+    Traces(TargetIter<TraceEvent>),
+}
+
+pub struct TargetIter<T> {
+    iter: std::vec::IntoIter<Value>,
+    metadata: EventMetadata,
+    _pahton: PhantomData<T>,
```
Comment (pablosichert) [inline lib/vector-core/src/event/vrl_target.rs:41]:
```suggestion
    _marker: PhantomData<T>,
```

File: lib/vector-core/src/event/log_event.rs [hunk 2]
```diff
@@ -149,6 +158,12 @@ impl LogEvent {
         Self { inner, metadata }
     }
 
+    ///  Create a `LogEvent` into a tuple of its components
```
Comment (pablosichert) [inline lib/vector-core/src/event/log_event.rs:161]:
```suggestion
    ///  Create a `LogEvent` into a tuple of its components.
```

File: lib/vector-core/src/event/vrl_target.rs [hunk 3]
```diff
@@ -55,19 +105,44 @@ impl VrlTarget {
...
-            ) as Box<dyn Iterator<Item = Event>>,
+            VrlTarget::LogEvent(value, metadata) => match value {
+                value @ Value::Object(_) => {
+                    TargetEvents::One(LogEvent::from_value(value, metadata).into())
+                }
+
+                Value::Array(values) => TargetEvents::Logs(TargetIter {
+                    iter: values.into_iter(),
+                    metadata,
+                    _pahton: PhantomData,
```
Comment (pablosichert) [inline lib/vector-core/src/event/vrl_target.rs:118]:
```suggestion
                    _marker: PhantomData,
```

File: lib/vector-core/src/event/vrl_target.rs [hunk 4]
```diff
@@ -55,19 +105,44 @@ impl VrlTarget {
...
+            VrlTarget::Trace(value, metadata) => match value {
+                value @ Value::Object(_) => {
+                    let log = LogEvent::from_value(value, metadata);
+                    TargetEvents::One(TraceEvent::from(log).into())
+                }
+
+                Value::Array(values) => TargetEvents::Traces(TargetIter {
+                    iter: values.into_iter(),
+                    metadata,
+                    _pahton: PhantomData,
```
Comment (pablosichert) [inline lib/vector-core/src/event/vrl_target.rs:136]:
```suggestion
                    _marker: PhantomData,
```

**Review by pablosichert — APPROVED** [review]
Looks good 👍

Some nits:

File: src/transforms/remap.rs [hunk 5]
```diff
@@ -427,11 +427,15 @@ where
...
-                    push_default(event, output, &self.default_schema_definition);
+            Ok(_) => match target.into_events() {
+                TargetEvents::One(event) => {
+                    push_default(event, output, &self.default_schema_definition)
                 }
-            }
+                TargetEvents::Logs(events) => events
+                    .for_each(|event| push_default(event, output, &self.default_schema_definition)),
+                TargetEvents::Traces(events) => events
+                    .for_each(|event| push_default(event, output, &self.default_schema_definition)),
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply src/transforms/remap.rs:437]: I looked into this, but it looks like we still need to iterate the individual events at the edge:

https://github.com/vectordotdev/vector/blob/980446d7efcbad039ac129d9740623382bed4728/lib/vector-core/src/transform/mod.rs#L422-L426

https://github.com/vectordotdev/vector/blob/980446d7efcbad039ac129d9740623382bed4728/lib/vector-core/src/transform/mod.rs#L398-L414

Unless you have thoughts on how we can avoid doing so?

**Pull request merged**